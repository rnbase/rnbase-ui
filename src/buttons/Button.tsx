import React from 'react';
import {
  Image,
  ImageStyle,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ImageSourcePropType,
  View,
  ViewStyle,
} from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';

export interface ButtonProps extends TouchableOpacityProps {
  children?: React.ReactNode;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  text?: React.ReactNode;
  textStyle?: StyleProp<TextStyle>;
  imageSource?: ImageSourcePropType;
  imageStyle?: StyleProp<ImageStyle>;
  imageAlignment?: 'left' | 'right';
}

export const Button: React.FC<Themed<typeof createStyleSheet, ButtonProps>> = ({
  styles,
  activeOpacity = 0.5,
  children,
  style,
  text,
  textStyle,
  imageSource,
  imageStyle,
  imageAlignment = 'left',
  ...props
}) => {
  let content;

  if (children) {
    content = children;
  } else {
    content = [];

    if (text) {
      const textStyles = [styles.text, textStyle, props.disabled && styles.disabledText];

      content.push(
        <Text key="text" style={textStyles} numberOfLines={1}>
          {text}
        </Text>,
      );
    }

    if (imageSource) {
      const imageStyles = [styles.image, imageStyle, props.disabled && styles.disabledImage];
      const image = <Image key="image" style={imageStyles} source={imageSource} />;

      if (imageAlignment === 'left') {
        content.unshift(image);
      } else {
        content.push(image);
      }
    }
  }

  const rootStyles = [styles.root, style, props.disabled && styles.disabledRoot];

  return (
    <TouchableOpacity activeOpacity={activeOpacity} {...props}>
      <View style={rootStyles}>{content}</View>
    </TouchableOpacity>
  );
};

export const createStyleSheet = ({ Colors, Fonts }: Theme) =>
  StyleSheet.create({
    root: {
      height: 50,
      borderRadius: 8,
      alignItems: 'center',
      flexDirection: 'row',
      paddingHorizontal: 15,
      justifyContent: 'center',
      backgroundColor: Colors.blue,
    },
    text: {
      ...Fonts.semibold,
      fontSize: 17,
      color: Colors.white,
      marginHorizontal: 5,
    },
    image: {
      width: 20,
      height: 20,
      flexShrink: 0,
      tintColor: Colors.white,
    },
    disabledRoot: {
      backgroundColor: Colors.gray5,
    },
    disabledText: {
      color: Colors.gray2,
    },
    disabledImage: {
      tintColor: Colors.gray2,
    },
  });

export type Props = ButtonProps & WithThemeProps;

export default withTheme(Button, createStyleSheet, 'Button');
