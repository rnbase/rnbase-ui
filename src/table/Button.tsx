import React from 'react';
import { StyleProp, TextStyle, ViewProps } from 'react-native';

import { Themed, withTheme } from '../theming';
import Row from './Row';

interface ButtonProps extends ViewProps {
  color?: string;
  height?: number;
  title: string;
  activeOpacity?: number;
  underlayColor?: string;
  titleStyle?: StyleProp<TextStyle>;
  onPress: () => void;
  onPressIn?: () => void;
  onPressOut?: () => void;
}

const Button: React.FC<Themed<typeof createStyleSheet, ButtonProps>> = ({
  theme: { colors },
  color = colors.blue,
  height,
  title,
  activeOpacity,
  underlayColor,
  style,
  titleStyle,
  onPress,
  onPressIn,
  onPressOut,
}) => (
  <Row
    height={height}
    title={title}
    activeOpacity={activeOpacity}
    underlayColor={underlayColor}
    disclosure={false}
    style={style}
    titleStyle={[{ color }, titleStyle]}
    onPress={onPress}
    onPressIn={onPressIn}
    onPressOut={onPressOut}
  />
);

const createStyleSheet = () => {};

export default withTheme(Button, createStyleSheet, 'Table.Button');
