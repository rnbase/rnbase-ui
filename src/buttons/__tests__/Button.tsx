import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

import { Props } from '../Button';

const TouchableOpacity = (props: any) => React.createElement('TouchableOpacity', props);

jest.doMock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => TouchableOpacity);

/**
 * Under test
 */
const { default: Button } = jest.requireActual('../Button');

const createElement = (props: Props) => <Button {...props} text="TEXT" />;

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render as disabled', () => {
  expect(createRenderer({ disabled: true })).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(
    createRenderer({
      style: { backgroundColor: '#000000' },
      textStyle: { color: '#ffffff' },
    }),
  ).toMatchSnapshot();
});

it('should render with icon', () => {
  expect(
    createRenderer({ imageSource: { uri: 'image.png' }, imageStyle: { tintColor: '#ffffff' } }),
  ).toMatchSnapshot();
});

it('should call onPress', () => {
  const onPress = jest.fn();
  const tree = createRenderer({ onPress });
  const button = tree.root.findByType(TouchableOpacity);

  button.props.onPress();

  expect(onPress).toBeCalled();
});
