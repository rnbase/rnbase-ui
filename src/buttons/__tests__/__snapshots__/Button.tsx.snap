// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as disabled 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={true}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#8e8e93",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        Object {
          "opacity": 0.5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render normally 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#8e8e93",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render with custom styles 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#8e8e93",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        Object {
          "backgroundColor": "#000000",
        },
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
  </View>
</TouchableOpacity>
`;

exports[`should render with icon 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#8e8e93",
          "borderRadius": 5,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 15,
        },
        undefined,
        false,
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 16,
            "height": 16,
            "lineHeight": 16,
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      <Text>
        TEXT
      </Text>
    </Text>
    <Image
      source={
        Object {
          "uri": "image.png",
        }
      }
      style={
        Array [
          Object {
            "flexShrink": 0,
            "height": 20,
            "tintColor": "#ffffff",
            "width": 20,
          },
          Object {
            "tintColor": "#ffffff",
          },
        ]
      }
    />
  </View>
</TouchableOpacity>
`;
